---
title: "prague_colaborators"
author: "RRPS"
date: '2022-07-14'
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(
	fig.height = 5,
	fig.width = 8,
	dpi = 180,
	message = FALSE,
	warning = FALSE,
	cache = TRUE,
	echo = TRUE
)
```

```{r CLEAR EVERYTHING, eval=FALSE, include=FALSE}
# unload all non-base packages
lapply(names(sessionInfo()$otherPkgs), function(pkgs)
  detach(
    paste0('package:', pkgs),
    character.only = T,
    unload = T,
    force = T
))

rm(list=ls())
```

```{r load libraries, message=FALSE, warning=FALSE, include=FALSE}
# load the packages
# main packs to start with
pcks <- c( "tidyverse", "janitor")
# require("tidyverse")

if(sum(as.numeric(!pcks %in% installed.packages())) != 0){
  installation <- pcks[!pcks %in% installed.packages()]
  for(i in 1:length(installation)) {
    install.packages(installation, dependencies = T)
    break()}
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
} else {
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
}

rm(pcks)
```

# setting working directories
```{r directories}
# get working directory and paste directory to save outputs
# in the Knit dropdown menu, select Knit directory, then Current Working Directory.
# dir.create("~/R/microbiome_br/scripts")
# dir.create("~/R/microbiome_br/outputs")
# dir.create("~/R/microbiome_br/data")
work_dir <- getwd()
folder_path <- "~/R/microbiome_br/outputs/prague/"
#folder_path <- "../../../microbiome_br_2/"
data.dir <- paste0(work_dir, "/data/prague")
# dir.create("data")
# dir.create(data.dir)
setwd(data.dir)

# # set colors
# color_distance <- c("D0" = "deepskyblue2", 'D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
# # color_distance1 <- c('D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
# color_depth <- c('DP1'="darkorange",'DP2'="purple", "DP3" = "cyan4")
# color_sites <- c(lease_1 = "#56B4E9", lease_2 ="#E69F00"  , Control =  "#009E73", other_leases = "#D1BCA8")

# check the working directory
paste("Working in directory:", getwd())
```

# DATA

if the status in "Complete Genome" < 8: get the biggest Status. if > 8: make 3 column with Q1, median, and Q3 and count the genomes.
If there isnÂ´t Complete Genome > Scaffold > Contig > Chromosome 

- `taxonomy` column = taxonomy cleaned from the `BEST_TAX` and `#Organism/Name` columns from Federica, euk and prok datasets
- `taxonomy_short` column = just Genus and Species (no strain, etc)
- `genus` column = Genus

## euk
```{r get the data}
# eukaryotic table
euk <- read_delim('eukaryotes.txt')
#euk <- read_delim('../../../microbiome_br_2/eukaryotes.txt')

euk.clean <- euk %>% 
  mutate(`#Organism/Name` = str_replace_all(`#Organism/Name`, "\\[|\\]","")) %>% 
  mutate(taxonomy = `#Organism/Name`, .after = `#Organism/Name`) %>%
  separate(col = taxonomy, into = c("Genus", "species"), sep = " ", remove = F) %>%
  unite("taxonomy_short",Genus:species, sep = " ", remove = F) %>%
  janitor::clean_names(.)
```

checking
```{r get the data}
euk %>% colnames()
euk.clean %>% colnames()

# check the number of taxonomies
euk.clean %>% count(taxonomy, sort = T)
euk.clean %>% count(taxonomy_short, sort = T)
euk.clean %>% count(genus, sort = T)

# take a look in the data
# euk.clean %>% filter(if_any(everything(), ~str_detect(., " = "))) %>% pull(organism)
# euk.clean %>% filter(if_any(everything(), ~str_detect(., " = ")))
```

## prok
```{r get the data}
# prokaryotic table
prok <- read_delim('prokaryotes.txt')
#prok <- read_delim('../../../microbiome_br_2/prokaryotes.txt')

prok.clean <- prok %>% 
  mutate(`#Organism/Name` = str_replace_all(`#Organism/Name`, "\\[|\\]","")) %>% 
  mutate(taxonomy = `#Organism/Name`, .after = `#Organism/Name`) %>%
  separate(col = taxonomy, into = c("Genus", "species"), sep = " ", remove = F) %>%
  unite("taxonomy_short",Genus:species, sep = " ", remove = F) %>%
  janitor::clean_names(.)
```

checking
```{r get the data}
# checking
prok.clean %>% count(taxonomy, sort = T)
prok.clean %>% count(taxonomy_short, sort = T)
prok.clean %>% count(genus, sort = T)
  
prok %>% colnames()
prok.clean %>% colnames()

#prok %>% slice(102225)

# lets take a look in the databases
prok.clean %>% 
  filter(if_any(everything(), ~str_detect(., "Koribacter versatilis Ellin345"))) %>% 
  filter(if_any(status, ~str_detect(., "Complete")))

# equal_sign_euk <- prok.clean %>% filter(if_any(everything(), ~str_detect(., "="))) %>% as.data.frame()
prok.clean %>% filter(if_any(everything(), ~str_detect(., "=")))
```

## Federica table

```{r get the data}
# table final
final.table <- read_delim('MG_FEDERICA_TABLE_FINAL.tab')
#final.table <- read_delim('../../../microbiome_br_2/MG_FEDERICA_TABLE_FINAL.tab')

# "\\[|\\]" regex removes any instance of '[' or ']'
table.clean <- final.table %>% 
  mutate(BEST_TAX = str_replace_all(BEST_TAX, "\\[|\\]","")) %>% 
  rename('taxonomy' = 'BEST_TAX') %>%
  separate(col = taxonomy, into = c("Genus", "species"), sep = " ", remove = F) %>%
  unite("taxonomy_short",Genus:species, sep = " ", remove = F) %>%
  clean_names()


write.table(x = table.clean, file = paste0(folder_path, "FINAL_TABLE_CLEAN_MG_FEDERICA.txt"), quote = F, sep = "\t", row.names = F)
```


```{r get the data}
# checking
table.clean %>% count(taxonomy, sort = T)
table.clean %>% count(taxonomy_short, sort = T)
table.clean %>% count(genus, sort = T)


table.clean %>% select(taxonomy) %>% filter(if_any(everything(), ~str_detect(., "bacterium S5")))
table.clean %>% distinct(taxonomy)
```

Get the taxonomies
```{r get the data}
# get the taxonomies from Federica table
taxonomies_federica_tax_long <- table.clean %>% 
                          filter(taxonomy != "-") %>% 
                          distinct(taxonomy) %>%
                          pull(taxonomy)

# as_tibble(taxonomies_federica_tax_long) %>% count(value, sort = T)

taxonomies_federica_tax_short <- table.clean %>% 
                          filter(taxonomy_short != "- NA") %>% 
                          distinct(taxonomy_short) %>%
                          pull(taxonomy_short)

# as_tibble(taxonomies_federica_tax_short) %>% count(value, sort = T)

taxonomies_federica_genus <- table.clean %>% 
                          filter(genus != "-") %>% 
                          distinct(genus) %>%
                          pull(genus)

# as_tibble(taxonomies_federica_genus) %>% count(value, sort = T)

```

# filtering

Here we are cleaning euk and prok datasets based on the taxonomies from Federica table. 

Filtering the euk and prok datasets based on the BEST_TAX column (Federica table), Genus species, and only Genus.

## prok 
```{r}
# filter prok table
prok.info <- prok.clean %>% 
              select(number_organism_name, taxonomy, taxonomy_short, genus, size_mb, gc_percent, status) %>% 
              mutate(gc_percent = as.numeric(gc_percent))

# BEST_TAX column Federica (former best_tax_prok)
best_tax_prok_tax_long <- taxonomies_federica_tax_long %>% map_df(., ~prok.info  %>% 
                                               filter(str_detect(taxonomy, .x)) %>% 
                                               mutate(taxonomy_long = .x)
                                       )

# Genus_species
best_tax_prok_tax_short <- taxonomies_federica_tax_short %>% map_df(., ~prok.info  %>%
                                               filter(str_detect(taxonomy_short, .x)) %>% 
                                               mutate(taxonomy_short = .x)
                                       )
# only Genus
best_tax_prok_genus <- taxonomies_federica_genus %>% map_df(., ~prok.info  %>% 
                                                              filter(str_detect(genus, .x)) %>%
                                                              mutate(taxonomy_genus = .x)
                                                            )

```

checking
```{r}
# some taxa were not detected in the prok.info datset
table.clean %>% select(taxonomy) %>% filter(if_any(everything(), ~str_detect(., 'Magnetococcus')))
prok.info %>% filter(if_any(everything(), ~str_detect(., 'Magnetococcus')))
best_tax_prok_genus %>% filter(if_any(everything(), ~str_detect(., 'Magnetococcus')))
#Check_names <- best_tax_prok %>% filter(if_any
# (everything(), ~str_detect(., 'Candidatus')))
```

Saving
```{r}
# Store hits found
write.table(x = best_tax_prok_tax_long, file = paste0(folder_path, "federica_found_taxonomy_prok_tax_long.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = best_tax_prok_tax_short, file = paste0(folder_path, "/federica_found_taxonomy_prok_tax_short.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = best_tax_prok_genus, file = paste0(folder_path, "/federica_found_taxonomy_prok_genus.txt"), quote = F, sep = "\t", row.names = F)

## read hits found. Avoid running all steps again.
# best_tax_prok <- read_delim(file = paste0(folder_path, "/data/federica_found_taxonomy_prok_tax_long.txt"))
# best_tax_prok <- read_delim(file = paste0(folder_path, "federica_found_taxonomy_prok.txt"))
best_tax_prok_tax_long <- read_delim(file = paste0(folder_path, "federica_found_taxonomy_prok.txt"))

best_tax_prok_tax_short <- read_delim(file = paste0(folder_path, "federica_found_taxonomy_prok_tax_short.txt"))

best_tax_prok_genus <- read_delim(file = paste0(folder_path, "federica_found_taxonomy_prok_genus.txt"))

# Checking for differences between organism name and Taxonomy from FEDERICA's table
#options(pillar.print_max = 10, pillar.print_min = 10, pillar.width = 35)

which(best_tax_prok$number_organism_name != best_tax_prok$taxonomy)
best_tax_prok[which(best_tax_prok$number_organism_name != best_tax_prok$taxonomy), c(1,6)]

waldo::compare(best_tax_prok$number_organism_name, best_tax_prok$taxonomy)

```


## euk

```{r}
# filter euk table
euk.info <- euk.clean %>% 
              select(number_organism_name, taxonomy, taxonomy_short, genus, size_mb, gc_percent, status) %>% 
              mutate(gc_percent = as.numeric(gc_percent))

# Indentify taxa without GC content info (353 found)
# euk.info[which(is.na(euk.info$gc_percent)),]
euk.info %>% filter(is.na(gc_percent))

# BEST_TAX column Federica (former best_tax_prok)
best_tax_euk_tax_long <- taxonomies_federica_tax_long %>% map_df(., ~euk.info  %>% 
                                               filter(str_detect(taxonomy, .x)) %>% 
                                               mutate(taxonomy_long = .x)
                                       )

# Genus_species
best_tax_euk_tax_short <- taxonomies_federica_tax_short %>% map_df(., ~euk.info  %>%
                                               filter(str_detect(taxonomy_short, .x)) %>% 
                                               mutate(taxonomy_short = .x)
                                       )
# only Genus
best_tax_euk_genus <- taxonomies_federica_genus %>% map_df(., ~euk.info  %>% 
                                               filter(str_detect(genus, .x)) %>% 
                                               mutate(taxonomy_genus = .x)
                                       )


```

Saving
```{r}
# Store euk hits found
write.table(x = best_tax_euk_tax_long, file = paste0(folder_path, "federica_found_taxonomy_euk_tax_long.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = best_tax_euk_tax_short, file = paste0(folder_path, "/federica_found_taxonomy_euk_tax_short.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = best_tax_euk_genus, file = paste0(folder_path, "/federica_found_taxonomy_euk_genus.txt"), quote = F, sep = "\t", row.names = F)

## Read hits found. Avoid running all steps again.
# best_tax_euk_tax_long <- read_delim(file = paste0(folder_path, "/federica_found_taxonomy_euk_tax_long.txt"))
# best_tax_euk_tax_short <- read_delim(file = paste0(folder_path, "/federica_found_taxonomy_euk_tax_short.txt"))
# best_tax_euk_genus <- read_delim(file = paste0(folder_path, "/federica_found_taxonomy_euk_genus.txt"))


# all.info <- euk.info %>% bind_rows(prok.info)
```

clean global environment
```{r}
rm(list=setdiff(ls(), c("folder_path","taxonomies_federica_tax_long", "taxonomies_federica_tax_short", "taxonomies_federica_genus", "best_tax_prok_tax_long","best_tax_prok_tax_short","best_tax_prok_genus", "best_tax_euk_tax_long","best_tax_euk_tax_short", 'best_tax_euk_genus', 'table.clean')))

```


Split the datasets based on status
For prokayotes

```{r}
# table_summ_prok <- best_tax_prok %>% 
#  select(taxonomy, status, size_mb, gc_percent) %>%
#   group_by(taxonomy, status) %>% 
#   skimr::skim(size_mb, gc_percent) %>%
#   left_join(best_tax_prok %>% 
#               group_by(taxonomy, status) %>% 
#               summarise(n_size_mb = n_distinct(size_mb),
#                         n_gc_percent = n_distinct(gc_percent))
#             )
# 
# table_summ_prok %>% head
# 
# table_summ_prok %>% distinct(skim_variable)
```

```{r}
# status
best_tax_prok_tax_long %>% distinct(status)
# best_tax_prok_complete_genome <- best_tax_prok %>% filter(status == 'Complete Genome')
# best_tax_prok_scaffold <- best_tax_prok %>% filter(status == 'Scaffold')
# best_tax_prok_contig <- best_tax_prok %>% filter(status == 'Contig')
# best_tax_prok_chromosome <- best_tax_prok %>% filter(status == 'Chromosome')

# best_tax_prok %>% summarise(mean_size_mb = mean(size_mb),
                    #           max_size_mb = max(size_mb),
                    #           min_size_mb = min(size_mb),
                    #           n_size_mb = n_distinct(size_mb)
                    #           ) %>%
  # filter(n_size_mb >= 3)
```


# Function for statistical summary

the columns `Taxonomy` and `Taxonomy_from_dataset` in the output are related to the taxonomy filtered in the prok and euk datasets by `BEST_TAX` column from Federica table

```{r}
## The function
#source(file = "~/R/microbiome_br/scripts/prague/alternative_function.R")
source(file = "alternative_function.R")

## Function usage:
## genomic_traits(taxa, best_tax_prok)
```


## prok
```{r}
# Create function to gather statistics about the genomic information, using the highest level of assembly and the longer assembly

number_of_entries_prok_tax_long <- best_tax_prok_tax_long %>% count(taxonomy, sort = T )
number_of_entries_prok_tax_short <- best_tax_prok_tax_short %>% count(taxonomy, sort = T )
number_of_entries_prok_genus <- best_tax_prok_genus %>% count(taxonomy, sort = T )

best_tax_prok_tax_short %>% distinct(taxonomy_short)

number_of_entries_prok_genus %>% filter(if_any(everything(), ~str_detect(., 'bacterium S5')))
# best_tax_prok %>% filter(if_any(everything(), ~str_detect(., 'bacterium S5')))
# 
# # See the taxa with the most entries
# number_of_entries

#number_of_entries[order(number_of_entries$n,decreasing = T),]

## Tests
# df = best_tax_prok 
# taxa = "Xanthomonas campestris pv. campestris" # 80
# taxa = "Cronobacter turicensis" # 16
# taxa = "Acetobacter pasteurianus IFO 3283-32" # 1
# taxa = "Acetobacter pasteurianus IFO 3283-01" # 2
# 
# match("Acetobacter pasteurianus IFO 3283-32", number_of_entries$taxonomy)
# number_of_entries[match(taxa, number_of_entries$taxonomy), "n"]
# test_entries <- number_of_entries[match(taxa, number_of_entries$taxonomy), "n"]

## Testing function
# taxa = "Acetobacter pasteurianus IFO 3283-01" 

# genomic_traits(taxa, best_tax_prok)

# check on the larger ones from the table
larger_ones_prok_tax_long <- number_of_entries_prok_tax_long %>% 
                  filter(n > 3) %>% 
                  pull(taxonomy)

larger_ones_prok_tax_short <- number_of_entries_prok_tax_short %>% 
                   filter(n > 300) %>% 
                  pull(taxonomy)

larger_ones_prok_genus <- number_of_entries_prok_genus %>% 
                  filter(n > 500) %>% 
                  pull(taxonomy)

# apply the function
prok_taxonomy_search_using_purrr_tax_long <- larger_ones_prok_tax_long %>% map_df(., ~genomic_traits(., best_tax_prok_tax_long))

prok_taxonomy_search_using_purrr_tax_short <- larger_ones_prok_tax_short %>% map_df(., ~genomic_traits(., best_tax_prok_tax_short))

prok_taxonomy_search_using_purrr_genus <- larger_ones_prok_genus %>% map_df(., ~genomic_traits(., best_tax_prok_genus))

prok_taxonomy_search_using_purrr %>% count(highest_assembly_level)

# taxonomy_search_using_apply <- lapply(X = number_of_entries$taxonomy[order(number_of_entries$n, decreasing = F)][1100:1163], FUN = genomic_traits)
# removed_list <- as_tibble(do.call(rbind,taxonomy_search_using_apply))

```

checking differences among the datasets
```{r}
prok_taxonomy_search_using_purrr_tax_long %>% 
  ggplot(aes(x = size_mb, y = gc_percent)) +
  geom_point(aes(color = highest_assembly_level))

prok_taxonomy_search_using_purrr_tax_short  %>% 
  ggplot(aes(x = size_mb, y = gc_percent)) +
  geom_point(aes(color = highest_assembly_level))

prok_taxonomy_search_using_purrr_genus  %>% 
  ggplot(aes(x = size_mb, y = gc_percent)) +
  geom_point(aes(color = highest_assembly_level))

prok_taxonomy_search_using_purrr_tax_long %>% select(where(is.numeric)) %>% psych::describe()
prok_taxonomy_search_using_purrr_tax_short %>% select(where(is.numeric)) %>% psych::describe()
prok_taxonomy_search_using_purrr_genus %>% select(where(is.numeric)) %>% psych::describe()
```

## euk
```{r}
# status
# best_tax_euk %>% distinct(status)
# table_summ_euk <- best_tax_euk %>% 
#  select(taxonomy, status, size_mb, gc_percent) %>%
#   group_by(taxonomy, status) %>% 
#   skimr::skim(size_mb, gc_percent) %>%
#   left_join(best_tax_euk %>% 
#               group_by(taxonomy, status) %>% 
#               summarise(n_size_mb = n_distinct(size_mb),
#                         n_gc_percent = n_distinct(gc_percent))
#             )
# 
# # table_summ_euk %>% distinct(skim_variable)
# # 
# # best_tax_euk %>% filter(taxonomy == "Alternaria alternata" & status == "Contig") %>%
# #   group_by(taxonomy, status) %>%
# #    summarise(n_size_mb = n_distinct(size_mb),
# #                         n_gc_percent = n_distinct(gc_percent)) 
# 
# best_tax_euk %>% 
#               group_by(taxonomy, status) %>% 
#               summarise(n_size_mb = n_distinct(size_mb),
#                         n_gc_percent = n_distinct(gc_percent)) %>%
#               filter(n_size_mb >= 3) 

```

Apply the function to the eukaryotic dataset
```{r}
# # See the taxa with the most entries
# number_of_entries
number_of_entries_euk_tax_long <- best_tax_euk_tax_long %>% count(taxonomy, sort = T )
number_of_entries_euk_tax_short <- best_tax_euk_tax_short %>% count(taxonomy, sort = T )
number_of_entries_euk_genus <- best_tax_euk_genus %>% count(taxonomy, sort = T )

best_tax_euk_tax_long %>% filter(taxonomy == "Homo sapiens")

# Testing function
# genomic_traits("Aaosphaeria arxii", best_tax_euk)

# check on the larger ones from the table
larger_ones_euk_tax_long <- number_of_entries_euk_tax_long %>% 
                  slice_head(n = 64) %>% 
                  pull(taxonomy)

larger_ones_euk_tax_short <- number_of_entries_euk_tax_short %>% 
                  slice_head(n = 64) %>% 
                  pull(taxonomy)

larger_ones_euk_genus <- number_of_entries_euk_genus %>% 
                  slice_head(n = 64) %>% 
                  pull(taxonomy)

# apply the function
euk_taxonomy_search_using_purrr_tax_long <- larger_ones_euk_tax_long %>% map_df(., ~genomic_traits(., best_tax_euk_tax_long))

euk_taxonomy_search_using_purrr_tax_short <- larger_ones_euk_tax_short %>% map_df(., ~genomic_traits(., best_tax_euk_tax_short))

best_tax_euk_tax_short %>% filter(taxonomy_short == "Danio Rerio")

euk_taxonomy_search_using_purrr_genus <- larger_ones_euk_genus %>% map_df(., ~genomic_traits(., best_tax_euk_genus))

euk_taxonomy_search_using_purrr %>% count(highest_assembly_level)
```

checking differences among the datasets

OBS: Homo sapiens, Mus musculus and Danio rerio are outliers. size_mb too big.
```{r}
too_big <- c("Danio rerio", "Homo sapiens", "Mus musculus")

euk_taxonomy_search_using_purrr_tax_long %>% 
  filter(!str_detect(Taxonomy, str_c(too_big, collapse =  "|"))) %>%
  ggplot(aes(x = size_mb, y = gc_percent)) +
  geom_point(aes(color = highest_assembly_level))

euk_taxonomy_search_using_purrr_tax_short  %>% 
  filter(!str_detect(Taxonomy, str_c(too_big, collapse =  "|"))) %>%
  ggplot(aes(x = size_mb, y = gc_percent)) +
  geom_point(aes(color = highest_assembly_level))

euk_taxonomy_search_using_purrr_genus  %>% 
  filter(!str_detect(Taxonomy, str_c(too_big, collapse =  "|"))) %>%
  ggplot(aes(x = size_mb, y = gc_percent)) +
  geom_point(aes(color = highest_assembly_level))

euk_taxonomy_search_using_purrr_tax_long %>% 
  filter(!str_detect(Taxonomy, str_c(too_big, collapse =  "|"))) %>%
  select(where(is.numeric)) %>% psych::describe()
euk_taxonomy_search_using_purrr_tax_short %>% 
  filter(!str_detect(Taxonomy, str_c(too_big, collapse =  "|"))) %>%
  select(where(is.numeric)) %>% psych::describe()
euk_taxonomy_search_using_purrr_genus %>% 
  filter(!str_detect(Taxonomy, str_c(too_big, collapse =  "|"))) %>%
  select(where(is.numeric)) %>% psych::describe()
```



Clean environment
```{r}
rm(list=setdiff(ls(), c("folder_path","taxonomies_federica_tax_long", "taxonomies_federica_tax_short", "taxonomies_federica_genus", "best_tax_prok_tax_long","best_tax_prok_tax_short","best_tax_prok_genus", "best_tax_euk_tax_long","best_tax_euk_tax_short", 'best_tax_euk_genus', 'table.clean', 'genomic_traits')))
```


# binding to Federica table

Add tax genome size and GC content to table Federica

## Prokaryotes
```{r}
number_of_entries_prok_tax_long <- best_tax_prok_tax_long %>% count(taxonomy, sort = T )
number_of_entries_prok_tax_short <- best_tax_prok_tax_short %>% count(taxonomy, sort = T )
number_of_entries_prok_genus <- best_tax_prok_genus %>% count(taxonomy, sort = T )

# only prokaryotes 
# full_dataset_prok2 <- table.clean %>%
#             filter(organism %in% best_tax_prok$taxonomy) %>%
#             bind_cols(map_df(.$organism, ~genomic_traits(., best_tax_prok)))

full_dataset_prok_tax_long <- number_of_entries_prok_tax_long %>% 
  pull(taxonomy) %>% 
  map_df(., genomic_traits, best_tax_prok_tax_long)

full_dataset_prok_tax_short <- number_of_entries_prok_tax_short %>%
  pull(taxonomy) %>%
  map_df(., genomic_traits, best_tax_prok_tax_short)

full_dataset_prok_genus <- number_of_entries_prok_genus %>% 
  pull(taxonomy) %>%
  map_df(., genomic_traits, best_tax_prok_genus)

# Store prok hits found
write.table(x = full_dataset_prok_tax_long, file = paste0(folder_path, "data/federica_full_prok_tax_long.txt"), quote = F, sep = "\t", row.names = F)
# full_dataset_prok_tax_long <- read_delim(file = paste0(folder_path, "data/federica_full_prok_tax_long.txt"))

write.table(x = full_dataset_prok_tax_short, file = paste0(folder_path, "data/federica_full_prok_tax_short.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = full_dataset_prok_genus, file = paste0(folder_path, "data/federica_full_prok_genus.txt"), quote = F, sep = "\t", row.names = F)
```

## Eukaryotes
```{r}
number_of_entries_euk_tax_long <- best_tax_euk_tax_long %>% count(taxonomy, sort = T )
number_of_entries_euk_tax_short <- best_tax_euk_tax_short %>% count(taxonomy, sort = T )
number_of_entries_euk_genus <- best_tax_euk_genus %>% count(taxonomy, sort = T )

# only eukaryotes 
# full_dataset_euk <- table.clean %>% 
#             filter(organism %in% best_tax_euk$taxonomy) %>%
#             bind_cols(map_df(.$organism, ~genomic_traits(., best_tax_euk)))

# full_dataset_euk_tax_long <- number_of_entries_euk_tax_long %>%
#                         pull(taxonomy) %>%
#                         map_df(., genomic_traits, best_tax_euk_tax_long)
full_dataset_euk_tax_long <- number_of_entries_euk_tax_long %>%
                        pull(taxonomy) %>%
                        map_df(., genomic_traits, best_tax_euk_tax_long)

full_dataset_euk_tax_short <- number_of_entries_euk_tax_short %>%
                        pull(taxonomy) %>%
                        map_df(., genomic_traits, best_tax_euk_tax_short)

full_dataset_euk_genus <- number_of_entries_euk_genus %>%
                        pull(taxonomy) %>%
                        map_df(., genomic_traits, best_tax_euk_genus)

# Store euk hits found
write.table(x = full_dataset_euk_tax_long, file = paste0(folder_path, "data/federica_full_euk_tax_long.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = full_dataset_euk_tax_short, file = paste0(folder_path, "data/federica_full_euk_tax_short.txt"), quote = F, sep = "\t", row.names = F)

write.table(x = full_dataset_euk_genus, file = paste0(folder_path, "data/federica_full_euk_genus.txt"), quote = F, sep = "\t", row.names = F)

```

Search and include lines at the whole table
```{r}
full_dataset <- full_dataset_euk %>% 
  select(-gc_percent_n_as) %>% 
  bind_rows(full_dataset_prok)

# Searching taxonomy from Federica_table into the full_datasets
## match(full_dataset$Taxonomy_from_dataset[1:2], table.clean$organism) # inverted match
## match(table.clean$organism[1:2], full_dataset$Taxonomy) # correct search

matched_Federica_table_to_full_dataset <- match(table.clean$taxonomy, full_dataset$Taxonomy)

# test_vec <- c(NA,1,2,3,NA,NA,3)
# which(is.na(test_vec))
# create_fill_table <- function(matched_info){
#   if(is.na(matched_info)){
#     NA
#   }
#   full_dataset[matched_info,]
# }
# create_fill_table(test_vec)

filler_for_final_table <- full_dataset[matched_Federica_table_to_full_dataset,]

#sum(is.na(filler_for_final_table$Taxonomy))

Federica_table_with_genomic_traits_info <- bind_cols(table.clean, filler_for_final_table)

write.table(x = Federica_table_with_genomic_traits_info, file = paste0(folder_path, "data/FEDERICA_MG_genomic_traits_added.txt"), quote = F, sep = "\t", row.names = F)

```

```{r}
# orgs_prok <- table.clean %>% 
#              filter(organism %in% best_tax_prok$taxonomy) %>%
#           # filter(organism %in% c("Desulfitobacterium hafniense DCB-2")) %>%
#            # distinct(organism) %>%
#             pull
# 
# full_dataset_prok1 <- table.clean %>% 
#                filter(organism %in% best_tax_prok$taxonomy) %>%
#               # filter(organism %in% c("Desulfitobacterium hafniense DCB-2")) %>%
#                bind_cols(map_df(orgs_prok, ~genomic_traits(., best_tax_prok)))


# prok_table_federica <- best_tax_prok %>% 
#   slice_head(n = 20) %>%
#   pull(taxonomy) %>%
#   map_df(., ~genomic_traits(., best_tax_prok)) %>%
#   rename(organism = Taxonomy) %>%
#   distinct() # remove duplicate rows
# 
# df_test_prok <- table.clean %>%
#   slice_head(n = 20) %>%
#   left_join(euk_table_federica) %>%
#   distinct()


# PS:
# couldn't find some taxa
# e.g., Magnetococcus sp. MC-1 (Federica table) -> Magnetococcus marinus MC-1 (prokayotic table)
# Pseudomonas syringae pv. phaseolicola 1448 (Federica table) -> there is any Pseudomonas syringae pv. phaseolicola 1448 in the prok table
# Symbiotaphrina kochii (Federica table) -> no Symbiotaphrina kochii in eukaryotic table
# 
# table.clean %>% select(number_organism_name) %>% filter(if_any(everything(), ~str_detect(., 'Symbiotaphrina')))
# euk.info %>% filter(if_any(everything(), ~str_detect(., 'Symbiotaphrina')))
# best_tax_euk %>% filter(if_any(everything(), ~str_detect(., 'Symbiotaphrina')))
```

```{r}

# orgs_euk <- table.clean %>% 
#              filter(organism %in% c("Aaosphaeria arxii", 'Abortiporus biennis', 'Absidia repens')) %>%
#              pull
# 
# full_dataset_euk <- table.clean %>% 
#             filter(organism %in% c("Aaosphaeria arxii", 'Abortiporus biennis', 'Absidia repens')) %>%
#             bind_cols(map_df(orgs_euk, ~genomic_traits(., best_tax_euk)))

# euk_table_federica <- best_tax_euk %>%
#   filter(taxonomy %in% c("Aaosphaeria arxii", 'Abortiporus biennis', 'Absidia repens')) %>%
#   pull(taxonomy) %>%
#   map_df(., ~genomic_traits(., best_tax_euk)) %>%
#   rename(organism = Taxonomy) %>%
#   distinct() # remove duplicate rows
# # 
# df_test_euk <- table.clean %>%
#    filter(organism %in% c("Aaosphaeria arxii", 'Abortiporus biennis', 'Absidia repens')) %>%
#  # slice_head(n = 20) %>%
#   left_join(euk_table_federica) %>%
#   distinct()


# df_test_euk %>% filter(organism == "Aaosphaeria arxii") %>% view
```

